src/ess/diffraction/filter.py

```diff
@@ -0,0 +1,124 @@
+     be changed using "minimum_threshold") it will be filtered out, if a proton charge
+     is 10% higher than the maximum (This won't actually make any difference in
+     comparison to just the max value) based on how Mantid did filtering of bad pulses.
+
+    :param da: sc.DataArray, the data array that is to be filtered.
+
+    :param proton_charge: sc.DataArray, the DataArray that contains the proton charges,
+     e.g. da.attrs["proton_charge"].value. Expected to have
+
+    :param minimum_threshold: float, the minimum threshold is a percentage used to
+     calculate the bad proton_charge values. It is used in this equation, (min =
+     mean(data) * minimum_threshold), the min is then used for filtering out what is
+     thought to be bad proton_charge values as a minimum in the filter.
+
+    :param maximum_threshold: float, the maximum threshold is a percentage used to
+     calculate the bad proton_charge values. It is used in this equation,
+     (max = max(data) * maximum_threshold), the max is then used for filtering out
+     what is though to be bad proton_charge values as a minimum in the filter.
+
+    :param data_time_coord: str, the name of the time coord in your DataArray, da,
+     defaults to "pulse_time"
+
+    :return: A DataArray that contains a dimension called "pulse_slices" in which "good"
+     data in slice 1, and "bad" data in slice "0". To access the good data you would
+     slice this return result like this: da["good_pulse", 1].
+    """
+    # Auto find the time dim name from attribute, assume 1D DataArray
+    proton_charge = proton_charge.transform_coords(
+        data_time_coord, {data_time_coord: proton_charge.dims[0]}, keep_inputs=False)
+
+    max_charge = sc.max(proton_charge.data)
+    max_charge *= maximum_threshold
+    min_charge = sc.mean(proton_charge.data)
+    min_charge *= minimum_threshold
+
+    good_pulse = (proton_charge >= min_charge) & (proton_charge <= max_charge)
+
+    return filter(da=da, condition_intervals=good_pulse, dim="good_pulse")
+
+
+def filter_by_value(da: sc.DataArray, min_value: sc.Variable, max_value: sc.Variable):
+    """
+    Return all data points between a minimum and maximum values of da in a DataArray
+
+    :param da: sc.DataArray, the sc.DataArray that is being filtered
+
+    :param min_value: sc.Variable, the lower bound of the filtering for the attribute
+
+    :param max_value: sc.Variable, the upper bound of the filtering for the attribute
+
+    :return: sc.DataArray or sc.Variable, whichever the attribute is in the original da
+     that is passed.
+    """
+    filter_coord_name = '_'.join(da.coords)
+    da.coords[filter_coord_name] = da.data
+    data_in_range = da.groupby(filter_coord_name,
+                               bins=sc.concatenate(min_value, max_value,
+                                                   filter_coord_name)).copy(0)
+    del data_in_range.coords[filter_coord_name]
+    del da.coords[filter_coord_name]
```

Comment:
When the `bins` argument is used, `groupby` accepts a variable as the first argument, so we can avoid the temporary insertion of a new coord:
```suggestion
    data_in_range = da.groupby(da.data,
                               bins=sc.concatenate(min_value, max_value,
                                                   filter_coord_name)).copy(0)
    del data_in_range.coords[filter_coord_name]
```

Reply:
Even with this change, it had the coord there, perhaps clean up isn't properly handled by groupby?

Reply:
I don't think `groupby` is adding coords to the input, can you double check?

Reply:
I just looked at it and it seems it was removing a coordinate that shouldn't be removed. So you are correct.